machine/machine.go

```diff
@@ -101,6 +105,67 @@ func GetMachineMemoryCapacity() (uint64, error) {
 	return memoryCapacity, err
 }
 
+// GetMachineMemoryByType returns information about memory capcity and number of DIMMs.
+// Information is retrieved from sysfs edac per-DIMM API (/sys/devices/system/edac/mc/)
+// introduced in kernel 3.6. Documentation can be found at
+// https://www.kernel.org/doc/Documentation/admin-guide/ras.rst.
+// Full list of memory types can be found in edac_mc.c
+// (https://github.com/torvalds/linux/blob/v5.5/drivers/edac/edac_mc.c#L198)
+func GetMachineMemoryByType(edacPath string) (map[string]*info.MemoryInfo, uint, error) {
+	var dimmCount uint
+	memory := map[string]*info.MemoryInfo{}
+	names, err := ioutil.ReadDir(edacPath)
+	// On some architectures (such as ARM) memory controller device may not exist.
+	// If this is the case then we ignore error and return empty slice.
+	_, ok := err.(*os.PathError)
+	if err != nil && ok {
+		return memory, dimmCount, nil
+	} else if err != nil {
+		return memory, dimmCount, err
+	}
+	isController := regexp.MustCompile("mc[0-9]+")
+	isDimm := regexp.MustCompile("dimm[0-9]+")
+	for _, controllerDir := range names {
+		controller := controllerDir.Name()
+		if isController.MatchString(controller) {
```

Comment:
nit: this is getting pretty deeply nested, which makes it hard to read.  One way we can reduce that is by changing
```golang
if isController.MatchString(controller) {
    ...
}
```
to 
```golang
if !isController.MatchString(controller) {
    continue
}
...
```
This applies to both regexp.